### ISSUE TITLE
Problem unit testing 

### ISSUE THREAD

**Issue by vernonr3** [goadesign/model#348]
I've been trying to add unit tests giving fairly complete coverage to /expr/views.go. I've hit a problem with f
func (dv *DeploymentView) AddElements(ehs ...ElementHolder) error {

It may be my lack of understanding of the way the software is supposed to work...
```go
func (dv *DeploymentView) AddElements(ehs ...ElementHolder) error {
	var nodes []*DeploymentNode
	for _, eh := range ehs {
		switch e := eh.(type) {
		case *DeploymentNode:
			if addDeploymentNodeChildren(dv, e) {
				nodes = append(nodes, e)
			}
		case *ContainerInstance:
			if dv.SoftwareSystemID == "" || dv.SoftwareSystemID == Registry[e.ContainerID].(*Container).System.ID {
				addElements(dv.ViewProps, e)
				nodes = append(nodes, e.Parent)
			}
		case *InfrastructureNode:
			addElements(dv.ViewProps, e)
			nodes = append(nodes, e.Parent)
		default:
			return fmt.Errorf("elements of type %T cannot be added to deployment views", eh)
		}
	}
.,.
}
```
The problem I'm seeing is in the case *ContainerInstance - line 440. The test asks whether dv.SoftwareSystemID == Registry[e.ContainerID] etc.
A good unit test would load Registry map with the e.ContainerID and then execute the AddElements function passing it the appropriate container.
However, the Registry map is created in code in expr/registry.go at line 54. This sets the key to idify(e.System.ID + ":" + e.Name). Since idify appears (if I understand it correctly) to use hashing ( a one way function) - and the function implies there will be a minimum of 2 keys in the map - I'm wondering how to work out which key refers to the container, which to the system etc.
I can understand why one wants unique keys - but why use a one way function? If a two way transform were used it would be much easier to work out the identity of the item in the registry? 
Have I missed the point / completely misunderstood?

**Comment by raphael** [comment]
Good question! Part of the answer is that the ID has to follow a certain format so it can work with the `structurizr` service. The idea is that the act of adding an element to the registry creates a unique ID. That ID cannot be "guessed" in the absolute (similar to adding a row to a database where the IDs would auto-increment). However to make the DSL work the ID needs to be consistent given properties that uniquely identify an element as per the C4 spec hence the hashing. 

So TL;DR one approach here could be for the test to create two container instances in different systems - one that the deployment view is for and one that it's not and then to make sure that only the container instance that corresponds to the right system gets added.

**Comment by vernonr3** [comment]
Thanks. I've worked out a way - Test done..
